File: Python/Medium/implement_queue_using_stacks.py
# Problem: Implement Queue using Stacks
# Link: https://leetcode.com/problems/implement-queue-using-stacks/
#
# Task:
# Implement a first in first out (FIFO) queue using only two stacks.
# The implemented queue should support all the standard queue operations:
# push(x), pop(), peek(), and empty().
#
# Example:
# Input:
# ["MyQueue", "push", "push", "peek", "pop", "empty"]
# [[], [1], [2], [], [], []]
# Output:
# [null, null, null, 1, 1, false]
#
# Explanation:
# MyQueue myQueue = new MyQueue();
# myQueue.push(1); // queue: [1]
# myQueue.push(2); // queue: [1, 2]
# myQueue.peek();  // return 1
# myQueue.pop();   // return 1
# myQueue.empty(); // return false


class MyQueue:
    def __init__(self):
        # input_stack: new elements pushed here
        # output_stack: elements popped/peeked from here
        self.input_stack = []
        self.output_stack = []

    # --------------------------------------------------------
    # push(x): add element to end of queue
    # --------------------------------------------------------
    def push(self, x: int) -> None:
        self.input_stack.append(x)

    # --------------------------------------------------------
    # pop(): remove element from front of queue
    # --------------------------------------------------------
    def pop(self) -> int:
        # if output stack is empty, move all from input to output
        if not self.output_stack:
            while self.input_stack:
                self.output_stack.append(self.input_stack.pop())
        # pop from output (which represents the queue front)
        return self.output_stack.pop()

    # --------------------------------------------------------
    # peek(): get the front element without removing
    # --------------------------------------------------------
    def peek(self) -> int:
        if not self.output_stack:
            while self.input_stack:
                self.output_stack.append(self.input_stack.pop())
        return self.output_stack[-1]  # top of output stack = front of queue

    # --------------------------------------------------------
    # empty(): check if queue is empty
    # --------------------------------------------------------
    def empty(self) -> bool:
        return not self.input_stack and not self.output_stack


# ------------------------------------------------------------
# Test Section
# ------------------------------------------------------------
if __name__ == "__main__":
    q = MyQueue()
    q.push(1)
    q.push(2)
    print(q.peek())  # 1
    print(q.pop())   # 1
    print(q.empty()) # False

File: CSharp/Medium/ImplementQueueUsingStacks.cs
// Problem: Implement Queue using Stacks
// Link: https://leetcode.com/problems/implement-queue-using-stacks/
//
// Task:
// Implement a first-in-first-out (FIFO) queue using only two stacks.
// The implemented queue should support standard operations:
// push(x), pop(), peek(), and empty().
//
// Example:
// Input:
// ["MyQueue", "push", "push", "peek", "pop", "empty"]
// [[], [1], [2], [], [], []]
// Output:
// [null, null, null, 1, 1, false]
//
// Explanation:
// MyQueue myQueue = new MyQueue();
// myQueue.Push(1); // queue: [1]
// myQueue.Push(2); // queue: [1, 2]
// myQueue.Peek();  // return 1
// myQueue.Pop();   // return 1
// myQueue.Empty(); // return false

using System;
using System.Collections.Generic;

public class MyQueue
{
    private Stack<int> inputStack;
    private Stack<int> outputStack;

    // ------------------------------------------------------------
    // Constructor
    // ------------------------------------------------------------
    public MyQueue()
    {
        inputStack = new Stack<int>();
        outputStack = new Stack<int>();
    }

    // ------------------------------------------------------------
    // Push(x): Add element to the end of the queue
    // ------------------------------------------------------------
    public void Push(int x)
    {
        inputStack.Push(x);
    }

    // ------------------------------------------------------------
    // Pop(): Remove element from the front of the queue
    // ------------------------------------------------------------
    public int Pop()
    {
        if (outputStack.Count == 0)
        {
            // Move all elements from inputStack to outputStack
            while (inputStack.Count > 0)
            {
                outputStack.Push(inputStack.Pop());
            }
        }

        return outputStack.Pop();
    }

    // ------------------------------------------------------------
    // Peek(): Get the front element without removing it
    // ------------------------------------------------------------
    public int Peek()
    {
        if (outputStack.Count == 0)
        {
            while (inputStack.Count > 0)
            {
                outputStack.Push(inputStack.Pop());
            }
        }

        return outputStack.Peek();
    }

    // ------------------------------------------------------------
    // Empty(): Check if the queue is empty
    // ------------------------------------------------------------
    public bool Empty()
    {
        return inputStack.Count == 0 && outputStack.Count == 0;
    }
}


// ------------------------------------------------------------
// Test Section
// ------------------------------------------------------------
public class Program
{
    public static void Main()
    {
        MyQueue myQueue = new MyQueue();

        myQueue.Push(1);
        myQueue.Push(2);

        Console.WriteLine("Peek: " + myQueue.Peek()); // 1
        Console.WriteLine("Pop: " + myQueue.Pop());   // 1
        Console.WriteLine("Empty: " + myQueue.Empty()); // False
    }
}
